pkg/elle/txn/cycle.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	CycleTypeNames = map[string]struct{}{
+		"G-nonadjacent-process":  {},
+		"G-nonadjacent-realtime": {},
+	}
+
+	for k, _ := range CycleAnomalySpecs {
+		CycleTypeNames[k] = struct{}{}
+	}
+
+	UnknownAnomalyTypes = map[string]struct{}{
+		"empty-transaction-graph": {},
+		"cycle-search-timeout":    {},
+	}
+
+	ProcessAnalysisTypes = map[string]struct{}{}
+	RealtimeAnalysisTypes = map[string]struct{}{}
+	for k, _ := range CycleTypeNames {
+		if strings.Contains(k, "process") {
+			ProcessAnalysisTypes[k] = struct{}{}
+		}
+		if strings.Contains(k, "realtime") {
+			RealtimeAnalysisTypes[k] = struct{}{}
+		}
+	}
+}
+
+// CycleExplainWrapper is a ICycleExplainer, it's also a wrapper for core.
+type CycleExplainerWrapper struct{}
+
+func (c CycleExplainerWrapper) ExplainCycle(pairExplainer core.DataExplainer, circle core.Circle) CycleCase {
+	panic("implement me")
+}
+
+func (c CycleExplainerWrapper) RenderCycleExplanation(explainer core.DataExplainer, circle core.Circle) string {
+	exp := core.CycleExplainer{}
+	return exp.RenderCycleExplanation(explainer, circle)
+}
+
+// NonadjacentRW is an strange helper function. It returns (valid, rw-count, current-is-rw).
+//  when initialize, please provide [0, true]. if you provide [0, true], please ensure that the first
+//  edge must not rw edge.
+// And if you provide [0, false], please finally call the first Rel again to check the tail-end rw adjacent.
+// This fn ensures that no :rw is next to another by testing successive edge
+//  types. In addition, we ensure that the first edge in the cycle is not an rw.
+//  Cycles must have at least two edges, and in order for no two rw edges to be
+//  adjacent, there must be at least one non-rw edge among them. This constraint
+//  ensures a sort of boundary condition for the first and last nodes--even if
+//  the last edge is rw, we don't have to worry about violating the nonadjacency
+//  property when we jump to the first.
+func NonadjacentRW(n int, lastIsRw bool, rel core.Rel) (bool, int, bool) {
+	if rel != core.RW {
+		return true, n, false
+	} else {
+		return !lastIsRw, n + 1, true
+	}
+}
+
+type AdditionalGraphOption struct {
+	ConsistencyModels []core.ConsistencyModelName
+	Anomalies         []string
+}
+
+type GraphFunc = func(core.History) (core.Anomalies, core.DirectedGraph, core.DataExplainer)
+
+// AdditionalGraphs determines what additional graphs we'll need to consider for this analysis.
+func AdditionalGraphs(opts AdditionalGraphOption, additionalGraphs []GraphFunc) []GraphFunc {
+	ats := reportableAnomalyTypes(opts.ConsistencyModels, opts.Anomalies)
+	var graphFn GraphFunc
+	if hasIntersection(ats, RealtimeAnalysisTypes) {
+		graphFn = core.RealtimeGraph
+	} else if hasIntersection(ats, ProcessAnalysisTypes) {
+		graphFn = core.ProcessGraph
+	} else {
+		return additionalGraphs
+	}
+	return append(additionalGraphs, graphFn)
```

Comment:
There is also monotonic key graph. Will there be multi additional graph?

Reply:
Yes, actually it will.

Reply:
Do we need to append multi graphs when this happens?

Reply:
In original code, `additionalGraphs` is part of `opts`, and it will append if it contains. Here I change it to array(because I don't like change `opts` in a helper function.)